File: sdk/python/examples/langgraph/39_tool_call_chain.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Tool Call Chain — chaining multiple tool calls in sequence.

Demonstrates:
    - An agent that calls several tools in a defined order
    - create_agent handles the ReAct loop automatically (no manual ToolNode/tools_condition)
    - State accumulation across multiple tool invocations
    - Practical use case: data enrichment pipeline (fetch → transform → validate → summarize)

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

import json

from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def fetch_company_info(company_name: str) -> str:
    """Look up basic information about a company."""
    data = {
        "openai": {"founded": 2015, "employees": "~1500", "sector": "AI Research"},
        "google": {"founded": 1998, "employees": "~190000", "sector": "Technology"},
        "microsoft": {"founded": 1975, "employees": "~220000", "sector": "Technology"},
        "anthropic": {"founded": 2021, "employees": "~500", "sector": "AI Safety"},
    }
    key = company_name.lower()
    if key in data:
        return json.dumps(data[key])
    return json.dumps({"error": f"Company '{company_name}' not found in database"})


@tool
def calculate_company_age(founded_year: int) -> str:
    """Calculate how many years a company has been in operation."""
    current_year = 2025
    age = current_year - founded_year
    return f"The company has been operating for {age} years (founded {founded_year})"


@tool
def get_sector_peers(sector: str) -> str:
    """Return a list of well-known companies in the same sector."""
    peers = {
        "ai research": ["OpenAI", "Anthropic", "DeepMind", "Cohere"],
        "ai safety": ["Anthropic", "OpenAI", "Redwood Research"],
        "technology": ["Apple", "Microsoft", "Google", "Meta", "Amazon"],
    }
    key = sector.lower()
    if key in peers:
        return f"Peers in '{sector}': {', '.join(peers[key])}"
    return f"No peer data available for sector: {sector}"


@tool
def generate_investment_note(company: str, age: str, peers: str) -> str:
    """Generate a brief investment note combining company facts."""
    return (
        f"Investment Note — {company}\n"
        f"Operational history: {age}\n"
        f"Competitive landscape: {peers}\n"
        f"Recommendation: Review financials and recent growth metrics before investing."
    )


graph = create_agent(
    llm,
    tools=[fetch_company_info, calculate_company_age, get_sector_peers, generate_investment_note],
    system_prompt=(
        "You are a financial analyst. For each company query, you MUST:\n"
        "1. Fetch company info\n"
        "2. Calculate company age using the founded year\n"
        "3. Get sector peers\n"
        "4. Generate an investment note combining all facts\n"
        "Call the tools in this order."
    ),
    name="tool_call_chain_agent",
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(graph, "Analyze Anthropic for investment purposes.")
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/39-tool-call-chain.ts
/**
 * Tool Call Chain -- chaining multiple tool calls in sequence.
 *
 * Demonstrates:
 *   - An agent that must call several tools in a defined order
 *   - Using ToolNode and tools_condition for standard LangGraph tool loop
 *   - State accumulation across multiple tool invocations
 *   - Practical use case: data enrichment pipeline (fetch -> transform -> validate)
 *
 * In production you would use:
 *   import { StateGraph, START, END } from '@langchain/langgraph';
 *   import { ToolNode, toolsCondition } from '@langchain/langgraph/prebuilt';
 */

import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// Tool implementations
// ---------------------------------------------------------------------------
const COMPANY_DATA: Record<string, { founded: number; employees: string; sector: string }> = {
  openai: { founded: 2015, employees: '~1500', sector: 'AI Research' },
  google: { founded: 1998, employees: '~190000', sector: 'Technology' },
  microsoft: { founded: 1975, employees: '~220000', sector: 'Technology' },
  anthropic: { founded: 2021, employees: '~500', sector: 'AI Safety' },
};

function fetchCompanyInfo(companyName: string): string {
  const data = COMPANY_DATA[companyName.toLowerCase()];
  if (data) return JSON.stringify(data);
  return JSON.stringify({ error: `Company '${companyName}' not found in database` });
}

function calculateCompanyAge(foundedYear: number): string {
  const age = 2025 - foundedYear;
  return `The company has been operating for ${age} years (founded ${foundedYear})`;
}

function getSectorPeers(sector: string): string {
  const peers: Record<string, string[]> = {
    'ai research': ['OpenAI', 'Anthropic', 'DeepMind', 'Cohere'],
    'ai safety': ['Anthropic', 'OpenAI', 'Redwood Research'],
    technology: ['Apple', 'Microsoft', 'Google', 'Meta', 'Amazon'],
  };
  const found = peers[sector.toLowerCase()];
  if (found) return `Peers in '${sector}': ${found.join(', ')}`;
  return `No peer data available for sector: ${sector}`;
}

function generateInvestmentNote(company: string, age: string, peers: string): string {
  return (
    `Investment Note -- ${company}\n` +
    `Operational history: ${age}\n` +
    `Competitive landscape: ${peers}\n` +
    `Recommendation: Review financials and recent growth metrics before investing.`
  );
}

// ---------------------------------------------------------------------------
// Mock compiled graph (simulates multi-step tool calls)
// ---------------------------------------------------------------------------
const graph = {
  name: 'tool_call_chain_agent',

  invoke: async (input: Record<string, unknown>) => {
    const query = (input.input as string) ?? '';

    // Extract company name from query
    const companyMatch = query.match(/\b(anthropic|openai|google|microsoft)\b/i);
    const company = companyMatch?.[1] ?? 'Anthropic';

    // Step 1: Fetch company info
    const info = fetchCompanyInfo(company);
    const parsed = JSON.parse(info);

    // Step 2: Calculate age
    const age = calculateCompanyAge(parsed.founded);

    // Step 3: Get sector peers
    const peers = getSectorPeers(parsed.sector);

    // Step 4: Generate investment note
    const note = generateInvestmentNote(company, age, peers);

    return {
      messages: [
        { role: 'user', content: query },
        { role: 'assistant', content: note },
      ],
    };
  },

  getGraph: () => ({
    nodes: new Map([
      ['__start__', {}],
      ['agent', {}],
      ['tools', {}],
      ['__end__', {}],
    ]),
    edges: [
      ['__start__', 'agent'],
      ['agent', 'tools'],
      ['tools', 'agent'],
      ['agent', '__end__'],
    ],
  }),

  nodes: new Map([
    ['agent', {}],
    ['tools', {}],
  ]),

  stream: async function* (input: Record<string, unknown>) {
    const query = (input.input as string) ?? '';
    const companyMatch = query.match(/\b(anthropic|openai|google|microsoft)\b/i);
    const company = companyMatch?.[1] ?? 'Anthropic';

    const info = fetchCompanyInfo(company);
    yield ['updates', { tools: { tool: 'fetch_company_info', result: info } }];

    const parsed = JSON.parse(info);
    const age = calculateCompanyAge(parsed.founded);
    yield ['updates', { tools: { tool: 'calculate_company_age', result: age } }];

    const peers = getSectorPeers(parsed.sector);
    yield ['updates', { tools: { tool: 'get_sector_peers', result: peers } }];

    const note = generateInvestmentNote(company, age, peers);
    yield ['updates', { agent: { messages: [{ role: 'assistant', content: note }] } }];
    yield ['values', { messages: [{ role: 'assistant', content: note }] }];
  },
};

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(
      graph,
      'Analyze Anthropic for investment purposes.',
    );
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
